File: c/p1382.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

static void traverse(struct TreeNode* root, int visited[], int* position) {
    if (!root) return;

    traverse(root->left, visited, position);
    visited[(*position)++] = root->val;
    traverse(root->right, visited, position);
}

static struct TreeNode* build_tree(struct TreeNode* tree, const int values[], const int left, const int right) {

    if (left >= right) return tree;

    const int length = right - left;
    const int l = left;
    const int m = l + length / 2;
    const int r = right;

    struct TreeNode* left_tree = build_tree(tree, values, l, m);
    struct TreeNode* right_tree = build_tree(tree, values, m + 1, r);

    struct TreeNode* n = calloc(1, sizeof(struct TreeNode));
    n->val = values[m];
    n->left = left_tree;
    n->right = right_tree;

    return n;
}


struct TreeNode* balanceBST(struct TreeNode* root) {
    // TODO: could save memory by pre-traversing to get size..
    // Linked list would slow down tree building..
    int visited[10000];
    int length = 0;
    traverse(root, visited, &length);

    return build_tree(NULL, visited, 0, length);
}

File: python3/p1382.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    @staticmethod
    def traverse(root: TreeNode, visited):
        if not root:
            return

        Solution.traverse(root.left, visited)
        visited.append(root.val)
        Solution.traverse(root.right, visited)


    @staticmethod
    def build_tree(tree: TreeNode, values: int) -> TreeNode:

        if not values:
            return tree

        left_values = values[:len(values) // 2]
        right_values = values[len(values) // 2:]
        middle_value = right_values.pop(0)

        left_tree = Solution.build_tree(tree, left_values)
        right_tree = Solution.build_tree(tree, right_values)

        return TreeNode(middle_value, left_tree, right_tree)


    def balanceBST(self, root: TreeNode) -> TreeNode:

        values = []
        Solution.traverse(root, values)

        return Solution.build_tree(None, values)
